### PR TITLE
Large block text setting

### LINKED ISSUE #2645: Bolder / larger symbols in math operator blocks
### Is your request related to a problem? Please describe.

The way that Scratch displays individual symbols in the math operator blocks can be pretty hard to read. For reference, here's what they look like:

image

### Describe the solution you'd like

It would be cool to have an extension that replaces these symbols with larger, bolder symbols. Certain ones, like the multiplication and division symbols, could be replaced with better substitutes **×** and **÷**.

### LINKED ISSUE #4913: Customizable block text style doesn't redraw blocks
### Description

When enabling or disabling customizable block text style or its "bold text" setting, the size of the blocks doesn't change. This results in weird spacing until the page is reloaded.

### Reproduction steps

1. Disable customizable block text style
2. Open the editor. Spacing is correct:
  image
4. Enable customizable block text style
5. Enable the "bold text" setting. The space around the text is too small:
  image
6. Reload. Spacing is correct:
  image
8. Disable the "bold text" setting. The space around the text is too large:
  image

### Expected behavior

A reload would not be required to adjust the width of blocks.

### PR DESCRIPTION
**Pull request by DNin01**
### Changes

Adds a percentage setting to the `custom-block-text` addon to increase the size of text on blocks up to 200%. You can also lower the size down to 75% in case for some reason you need to.

Image

### Reason for changes

To make the blocks easier to read.

### Tests

Tested in Edge 114 in combination with `middle-click-popup`.

Resolves #4913

- [x] [TODOs](https://github.com/ScratchAddons/ScratchAddons/pull/5917#issuecomment-1616180035)

### CODE REVIEW

**Comment by mxmou** [comment]
I don't think this should close #2645. I agree with [this comment](https://github.com/ScratchAddons/ScratchAddons/issues/2645#issuecomment-893058277):

> Some people don't have issues with the general font of the editor, only the arithmetic blocks since they have such small symbols.

**Comment by BroJac5246** [comment]
> I don't think this should close #2645. I agree with [this comment](https://github.com/ScratchAddons/ScratchAddons/issues/2645#issuecomment-893058277):
> 
> > Some people don't have issues with the general font of the editor, only the arithmetic blocks since they have such small symbols.

Could there be separate settings for text and operators/symbols?

**Comment by Secret-chest** [comment]
Why no settings?

**Comment by WorldLanguages** [comment]
I also wonder why you've discarded the idea of making the font size customizable.

**Comment by DNin01** [comment]
> Why no settings? (font size)
> 
> They're really easy to implement.

> I also wonder why you've discarded the idea of making the font size customizable.

Alright, so if I make the size customizable, should I use an integer input or a couple single-select size options like "normal", "large" and "larger"?

**Comment by WorldLanguages** [comment]
@DNin01 We can combine both concepts. I'd prefer having a select with options default/large/larger/custom. If user chose "custom" they could manually enter an integer value.

This integer value doesn't necessarily need to match one-to-one to CSS units. Ideally we should allow users to change the font one step at a time as defined by browser behavior. For example, I don't think most browsers distinguish 14.2pt from 14.3pt. Maybe browsers actually round it to the nearest integer and I've written all of this for no reason? Do we need to read the CSS spec?

**Comment by DNin01** [comment]
I have added the option to choose between 5 different larger text sizes.

**Comment by mxmou** [comment]
I think it's better to just make the font size fully customizable (with reasonable minimum and maximum values). A checkbox that enables a number or text setting is sometimes useful, but in this case it just adds unnecessary complexity.

**Comment by Secret-chest** [comment]
> I think it's better to just make the font size fully customizable (with reasonable minimum and maximum values). A checkbox that enables a number or text setting is sometimes useful, but in this case it just adds unnecessary complexity.

Completely agree. It should be a multiplier, maybe 75% to 250%.

**Comment by Secret-chest** [comment]
Since it's a percentage, you can also tell how large it's going to be.

**Comment by Secret-chest** [comment]
> Maybe browsers actually round it to the nearest integer and I've written all of this for no reason? Do we need to read the CSS spec?

It could depend on the scaling / zoom. And even on a normal 1920x1080px screen with 100% zoom, `13px` and `13.5px` look different.

**Comment by WorldLanguages** [comment]
> I think it's better to just make the font size fully customizable (with reasonable minimum and maximum values). A checkbox that enables a number or text setting is sometimes useful, but in this case it just adds unnecessary complexity.

My concern was that the settings page currently doesn't have a per-setting reset button if the addon has no presets. Unless we use percentages, we wouldn't really expect the user to know or remember what the default text size is.
This might cause some confusion:
1. The user might think the addon defaults to smaller or larger text. How would they know if vanilla Scratch's font size is 3, 13, or whatever? We know that this addon doesn't change anything by default, but will users know?
2.  The user might want to return to the default text size (after playing with the setting) and won't remember what the vanilla Scratch value was. So their only option is to reset the whole addon to default settings.

**Comment by DNin01** [comment]
> > I think it's better to just make the font size fully customizable (with reasonable minimum and maximum values). A checkbox that enables a number or text setting is sometimes useful, but in this case it just adds unnecessary complexity.
> 
> Completely agree. It should be a multiplier, maybe 75% to 250%.
> This would also remove the need for a switch, since 100% would be unchanged.

Interesting 🤔

**Comment by DNin01** [comment]
> 1. The user might think the addon defaults to smaller or larger text. How would they know if vanilla Scratch's font size is 3, 13, or whatever? We know that this addon doesn't change anything by default, but will users know?

Turning off the "large text" setting would revert the text size back to the default.

**Comment by Secret-chest** [comment]
Still, why have 5 levels? A percentage would be simpler to understand.

**Comment by DNin01** [comment]
Yeah, let's try percentages.

**Comment by DNin01** [comment]
It might look nicer if the dropdown arrows changed size too.

**Comment by Secret-chest** [comment]
You can have the minimum and default be 100, then remove the toggle.

**Comment by mxmou** [comment]
> You can have the minimum and default be 100, then remove the toggle.

The minimum should probably be less than 100. I don't think there's a reason to not let people make the font smaller.

**Comment by WorldLanguages** [comment]
> Turning off the "large text" setting would revert the text size back to the default.

Note that if the default font size setting is 100, the "large text" toggle would become unnecessary. I only proposed a toggle so that users understand that the default font size of the addon is exactly equal to vanilla Scratch's font size, but that would become clear if the number 100 meant "same as vanilla Scratch".

**Comment by DNin01** [comment]
Removed the toggle. For no change, text size can be set to 100% but it can go up to 150%.

**Comment by Secret-chest** [comment]
Great. You can make the limits 50 to 200%.

**Comment by DNin01** [comment]
> Great. You can make the limits 50 to 200%.

I'm still unsure why people would want to _decrease_ the text size below the already small default size, unless they typically zoom in. I could also increase the maximum but the text might not fit on the blocks.

I'll keep thinking about it. I suppose it wouldn't hurt to have the option, right?

**Comment by BroJac5246** [comment]
> > Great. You can make the limits 50 to 200%.
> 
> I'm still unsure why people would want to _decrease_ the text size below the already small default size, unless they typically zoom in. I could also increase the maximum but the text might not fit on the blocks unless people use a different addon to make them even bigger.
> 
> I'll keep thinking about it. I suppose it wouldn't hurt to have the option, right?

I'm sure there are reasons (some people just find things too large), and it doesn't hurt

File: addons/custom-block-text/addon.json [hunk 1]
```diff
@@ -36,11 +44,23 @@
...
+      "max": 150,
+      "default": 100
     }
   ],
   "dynamicEnable": true,
   "dynamicDisable": true,
   "updateUserstylesOnSettingsChange": true,
   "versionAdded": "1.24.0",
+  "latestUpdate": {
+    "version": "1.32.0",
```
Comment (WorldLanguages) [inline addons/custom-block-text/addon.json:62]:
```suggestion
    "version": "1.33.0",
```

**Comment by WorldLanguages** [comment]
Unless decreasing the font size leads to visual bugs, I think we should offer the option.

**Comment by WorldLanguages** [comment]
Is it just me, or the size of the blocks is determined only on first editor load? Dynamically changing the value of the setting results in this, and I can't find a way to fix it without a reload:

image

In the screenshot I opened the editor, then changed the setting from 100 to 150 by typing in the 3 digits.

**Comment by DNin01** [comment]
> Is it just me, or the size of the blocks is determined only on first editor load? Dynamically changing the value of the setting results in this, and I can't find a way to fix it without a reload:
> 
> In the screenshot I opened the editor, then changed the setting from 100 to 150 by typing in the 3 digits.

Looks like this bug: #4913

**Comment by WorldLanguages** [comment]
If we don't find a way to fix it, we can tell users that a reload is required for the settings to apply.
The bug wasn't that noticeable with the bold text.

**Comment by DNin01** [comment]
@mxmou Any chance you know how to fix this bug?

**Comment by WorldLanguages** [comment]
> If we don't find a way to fix it, we can tell users that a reload is required for the settings to apply.

This might not be good enough - it looks like there's some code (probably from Blockly) which at some specific point of time calculates the size of the blocks based on the font size. If our CSS, for some reason, gets injected after that moment, the bug will be visible to the user even though they haven't changed the settings dynamically.

It would be nice to know which Scratch code is responsible for this. We might try to find a way to call it multiple times, or at least being able to detect if it has ran already, to avoid race condition bugs.

**Comment by WorldLanguages** [comment]
Thanks @towerofnix for the help! The bug is now fixed.

**Comment by DNin01** [comment]
The block widths should be updated when the addon is dynamically disabled or when "bold text" is toggled. Also, after a second dynamic enable, it seems the text size isn't re-applied until a setting is changed.

**Comment by WorldLanguages** [comment]
> when "bold text" is toggled

Yeah, handling that separately is causing its own set of issues because even though the font size did adjust properly, it looks wrong because it adjusted before the bold text had a chance to apply.

Will fix these tomorrow

**Comment by DNin01** [comment]
Yes! It works great!

But I just realized that the blocks inside the Insert blocks by name addon's popup still require a page refresh to update.

**Comment by WorldLanguages** [comment]
> But I just realized that the blocks inside the Insert blocks by name addon's popup still require a page refresh to update.

Forgot about that addon!

We should also test changing the font size while the "make a block" modal is open.

**Comment by WorldLanguages** [comment]
Also, looks like we're currently styling the "make a X" buttons as if they were blocks:

image

**Comment by DNin01** [comment]
> Also, looks like we're currently styling the "make a X" buttons as if they were blocks:
> 
> image

Interesting. Maybe we could turn that into a feature?

**Comment by Tacodiva** [comment]
Just pushed fixes the issues with the middle click popup :blush:

File: addons/custom-block-text/userscript.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  textShadowCss.disabled = true;
+  document.head.appendChild(textShadowCss);
+
+  const updateBlockly = () => {
+    blocklyInstance.Field.cacheWidths_ = {}; // Clear text width cache
+    updateAllBlocks(vm);
+  };
+
+  const setFontSize = (wantedSize) => {
+    if (wantedSize !== 100) document.documentElement.style.setProperty("--customBlockText-sizeSetting", wantedSize);
```
Comment (Tacodiva) [inline addons/custom-block-text/userscript.js:47]: Nitpick, but could you not use the "customCssVariables" key in `addon.json` to do this automatically?

Comment (WorldLanguages) [reply addons/custom-block-text/userscript.js:47]: Yeah, but everything here is intentionally handled only by the userscript to make timing 0% dependent on any code outside this addon.
This addon doesn't use userstyles either for this same reason.

**Comment by WorldLanguages** [comment]
TODOs:

- [ ] This affects the "make a block" modal. Should it? Is it easy to avoid it? Also, test changing the font size while this the modal is being shown, and right after closing the modal.
- [ ] "Make a Variable", "Make a List" and "Make a Block" ideally shouldn't be styled, is it easy to exclude them?
- [ ] This addon could benefit from #4919 as it interacts with Blockly workspaces.

**Comment by CST1229** [comment]
> Question: is there really a difference between this setting to increase the font size, and zooming into the workspace? There really isn't right? That's a bit confusing.

Increasing the font size doesn't increase outline and padding sizes.

**Comment by DNin01** [comment]
> Question: is there really a difference between this setting to increase the font size, and zooming into the workspace? There really isn't right? That's a bit confusing.

You can't change the zoom level of the blocks palette, so this is the only way to make that text larger, other than changing the entire page's zoom level.

**Comment by DNin01** [comment]
Interesting issue: when pulling up the Make a Block popup, the blocks on the blocks palette suddenly stop adapting to text style changes, but everything else adapts.

To reproduce:
1. Click "Make a Block"
2. Change the text size

The block in the Make a Block menu itself also doesn't adapt.

**Comment by WorldLanguages** [comment]
> Interesting issue: when pulling up the Make a Block popup, the blocks on the blocks palette suddenly stop adapting to text style changes, but everything else continues to adapt.

Yeah, #6227 should fix that

**Comment by DNin01** [comment]
@WorldLanguages Apparently this branch also [got on the ScratchAddons repo](https://github.com/ScratchAddons/ScratchAddons/activity?ref=large-block-text-setting)? Noticed while I was trying to merge `ScratchAddons:master` into the branch from the web UI. Can we safety delete it?

**Comment by WorldLanguages** [comment]
I'm not sure, we can delete it after merging this

**Comment by DNin01** [comment]
> I'm not sure, we can delete it after merging this

It looks to me like we won't need it. We can restore it if deleting it causes issues.

**Comment by DNin01** [comment]
Problem solved!

**Review by Samq64 — COMMENTED** [review]
Could the style elements be moved to one or more CSS files?

**Comment by WorldLanguages** [comment]
> Could the style elements be moved to one or more CSS files?

@Samq64 I'm not sure if that can be done without sacrificing reliability, as loading CSS files from disk may take a few milliseconds, leading to race conditions.

**Comment by WorldLanguages** [comment]
> > Interesting issue: when pulling up the Make a Block popup, the blocks on the blocks palette suddenly stop adapting to text style changes, but everything else continues to adapt.
> 
> Yeah, #6227 should fix that

The PR was postponed to v1.35.0 milestone, so this PR should also be postponed

**Comment by WorldLanguages** [comment]
> Interesting issue: when pulling up the Make a Block popup, the blocks on the blocks palette suddenly stop adapting to text style changes, but everything else continues to adapt.
> 
> To reproduce:
> 
> 1. Click "Make a Block"
> 2. Change the text size
> 
> The block in the Make a Block menu itself also doesn't adapt.

Fixed!

File: addons/custom-block-text/addon.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "Customizable block text style",
-  "description": "Changes the thickness of the text on blocks and optionally adds a text shadow.",
+  "description": "Allows you to make a few adjustments to block text to make it easier to read.",
```
Comment (WorldLanguages) [inline addons/custom-block-text/addon.json:3]: Maybe add the keywords thickness, bold, and/or size to the description?

File: addons/custom-block-text/userscript.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  const blocklyInstance = await addon.tab.traps.getBlockly();
+
+  // Handling the CSS from here instead of a userstyle is much more stable, as
+  // there's no code outside of this addon dynamically toggling the styles.
+  // This way, we can clearly control the execution order of style operations.
+  // For example, we always want to call updateAllBlocks() after the styles
+  // were updated according to the user's settings, not before.
+  const fontSizeCss = document.createElement("style");
+  fontSizeCss.textContent = `
+    .blocklyText,
```
Comment (WorldLanguages) [inline addons/custom-block-text/userscript.js:17]:
```suggestion
    .blocklyText:not(.blocklyFlyoutButton > .blocklyText),
```

This change would make the CSS not affect the font size of the "make a variable", "make a block", etc. buttons.

image

**Comment by WorldLanguages** [comment]
Should this addon and `custom-block-shape` affect the "Make new block" block preview?
Currently, they do. It doesn't cause any issues unless you change the settings while the "new block" modal is open. In that case, you need to close and reopen it for it to readjust. No other known issues or incompatibilities with other addons.

**Comment by WorldLanguages** [comment]
We could also attempt to make the block category labels bigger/smaller depending on the block size.
Currently there's no way to change their size.
If we decide to change these, the "Make a Variable" and other buttons would be affected too.

This is how it looks right now: (font size 175%)
image

**Comment by mxmou** [comment]
> Should this addon and `custom-block-shape` affect the "Make new block" block preview?

I think so.

**Comment by WorldLanguages** [comment]
> We could also attempt to make the block category labels bigger/smaller depending on the block size.
> Currently there's no way to change their size.
> If we decide to change these, the "Make a Variable" and other buttons would be affected too.

I think we should also increase the font size of those, and merge this PR.

File: addons/custom-block-text/userscript.js (hunk 4 above)
Comment (WorldLanguages) [reply addons/custom-block-text/userscript.js:19]: I think we decided against this.

**Review by WorldLanguages — APPROVED** [review]

**Review by mxmou — APPROVED** [review]

**Comment by WorldLanguages** [comment]
From feedback: custom font size for editor comments. Seems like the missing piece of the puzzle. It's the only UI element in the Code tab that we don't provide a custom font size setting for.

**Pull request merged**